### PR TITLE
Support for floating point

### LINKED ISSUE #448: Add support for floating point types
These could almost be extern types, but we probably want support for parsing double/float literals.

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Fixes #448 
This PR should not yet be merged - we have first to fix the way fp numbers are printed. This is related to #537. But it can be reviewed.

### CODE REVIEW

File: test/datalog_tests/simple.dump.expected [hunk 1]
```diff
@@ -287,22 +342,53 @@ BigintVectors{"filter_C0(C0{1}, C1{2}, C1{0}, C1{3})",[2]}
 BigintVectors{"filter_gt((10, 20, 30), 15)",[20,30]}
 
 Cast_bigint:
-Cast_bigint{"(32'sd0 - 32'sd10) as bigint",-10,-10}
```
Comment (ryzhyk) [inline test/datalog_tests/simple.dump.expected]: why did this have to go away?

File: lib/libfloat.dl [hunk 2]
```diff
@@ -0,0 +1,86 @@
+// Library with operations on floating point values (float and double)
+// All float operations have suffix _f, and all double operations have suffix _d
+
+extern function floor_f(f: float): float
```
Comment (ryzhyk) [inline lib/libfloat.dl]: No tests for any of these functions? There's a script for testing libraries now (see `test/datalog_tests/test-libs.sh`)

File: lib/libfloat.dl [hunk 3]
```diff
@@ -0,0 +1,86 @@
+// Library with operations on floating point values (float and double)
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply lib/libfloat.dl]: float is a reserved keyword, but we can call it fp

[the remark this answers is not in the document]
Comment (ryzhyk) [reply lib/libfloat.dl]: `fp` sounds good

File: rust/template/differential_datalog/int.rs [hunk 4]
```diff
@@ -66,6 +68,16 @@ impl Int {
     pub fn from_i128(v: i128) -> Int {
         Int { x: BigInt::from(v) }
     }
+    pub fn from_float(v: OrderedFloat<f32>) -> Int {
+        Int {
+            x: BigInt::from_f32(*v).unwrap(),
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply rust/template/differential_datalog/int.rs]: I have kept the same APIs. I think this is less confusing.
I can probably make it NaN if from_f32 fails.

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply rust/template/differential_datalog/int.rs]: Sorry, there is no NaN in BigInt...

File: src/Language/DifferentialDatalog/Compile.hs [hunk 5]
```diff
@@ -2423,40 +2487,46 @@ mkBinOp d op (e1, t1) (e2, t2) =
...
                | otherwise
                -> parens $ e1 <+> "<<" <+> e2
         BAnd   -> parens $ e1 <+> "&"  <+> e2
         BOr    -> parens $ e1 <+> "|"  <+> e2
         BXor   -> parens $ e1 <+> "^"  <+> e2
-        Plus   | uint
+        Plus   | smallInt d t1
                -> parens $ e1 <> ".wrapping_add(" <> e2 <> ")"
+               | isFP d t1
+               -> "OrderedFloat" <> (parens $ e1 <> ".into_inner()" <+> "+" <+> e2 <> ".into_inner()")
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply src/Language/DifferentialDatalog/Compile.hs]:
```
error[E0369]: binary operation `+` cannot be applied to type `ordered_float::OrderedFloat<f64>`
d: OrderedFloat(OrderedFloat::<f64>(5.0) + OrderedFloat::<f64>(3.0))
 = note: an implementation of `std::ops::Add` might be missing for `ordered_float::OrderedFloat<f64>`
```

[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Compile.hs]: Ok. I wonder if there's a reason OrderedFloat does not implement `Add`.

File: src/Language/DifferentialDatalog/Validate.hs [hunk 6]
```diff
@@ -732,6 +738,8 @@ exprInjectStringConversions d ctx e@(EBinOp p Concat l r) | (te == tString) && (
                   TString{}   -> return $ bUILTIN_2STRING_FUNC
                   TBit{}      -> return $ bUILTIN_2STRING_FUNC
                   TSigned{}   -> return $ bUILTIN_2STRING_FUNC
+                  TDouble{}   -> return $ bUILTIN_2STRING_FUNC
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply src/Language/DifferentialDatalog/Validate.hs]: I am using now string interpolation with a double, so hopefully that will qualify

**Comment by mbudiu-vmw** [comment]
This should be ready for merging - if tests pass.

**Pull request closed without merging**